### PR TITLE
Support flashing .hex files onto any suitable External flash partition.

### PR DESCRIPTION
**Pull request by hydra**
This allows uploading not only firmware, but system, osd logos/fonts and config backups too.
Additionally, the firmware partition is not always the 3rd partition on new FCs, other partitions can preceed it.

A partition is 'suitable' if it matches the length of the .hex file.

### CODE REVIEW

**Review by haslinghuis — CHANGES_REQUESTED** [review]
Please solve the code smells too.

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - Github identifies PR as mergeable -> FAIL
 - PR is not in draft -> PASS
 - Assigned to a milestone -> PASS
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - 'Don't merge' label NOT found -> PASS
 - At least one 'RN:' label found -> PASS
 - 'Tested' label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

File: src/js/protocols/stm32usbdfu.js
Comment (haslinghuis) [inline src/js/protocols/stm32usbdfu.js:337]:
```suggestion
            // Descriptors:
```

File: src/js/protocols/stm32usbdfu.js
Comment (haslinghuis) [inline src/js/protocols/stm32usbdfu.js:606]:
```suggestion
                        for (let partitionIndex = 0; partitionIndex < self.flash_layout.sectors.length; partitionIndex++) {
```

File: src/js/protocols/stm32usbdfu.js
Comment (haslinghuis) [inline src/js/protocols/stm32usbdfu.js:608]:
```suggestion
                            const firmware_sectors = self.flash_layout.sectors[partitionIndex];
```

File: src/js/protocols/stm32usbdfu.js
Comment (haslinghuis) [inline src/js/protocols/stm32usbdfu.js:609]:
```suggestion
                            const firmware_partition_size = firmware_sectors.total_size;
```

File: src/js/protocols/stm32usbdfu.js
Comment (haslinghuis) [inline src/js/protocols/stm32usbdfu.js:611]:
```suggestion
                            if (firmware_partition_size === self.hex.bytes_total) {
```

File: src/js/protocols/stm32usbdfu.js
Comment (haslinghuis) [inline src/js/protocols/stm32usbdfu.js:619]:
```suggestion
                        if (self.available_flash_size === 0) {
```

**Review by haslinghuis — CHANGES_REQUESTED** [review]
@hydra you forgot some code smells, I have pointed them out to make it easier.

**Comment by hydra** [comment]
@haslinghuis I hadn't looked at the code-smells yet, the previous push was fixing the build failure.  My IDE doesn't have an option for removing trailing spacing on Javascript files (it does on C/JAVA files...)  Missed one, just forced pushed again to fix the other issues.

**Review by haslinghuis — APPROVED** [review]

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - Github identifies PR as mergeable -> FAIL
 - PR is not in draft -> PASS
 - Assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - 'Don't merge' label NOT found -> PASS
 - At least one 'RN:' label found -> PASS
 - 'Tested' label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

**Comment by haslinghuis** [comment]
@hydra which hardware is needed to test this - the merging is automated - and we need to test the PR before we can apply the `tested` label.

**Comment by hydra** [comment]
@haslinghuis SPRacingH7 EXTREME/ZERO/NANO/RF/EXTREME-PX4-EDITION/CINE

**Comment by Asizon** [comment]
@hydra these advantages looks so good, can you suply to our Configurator Maintainers with some hardware samples to can test all your new features?

**Comment by hydra** [comment]
Pretty sure the maintainers already have the SPRacingH7EXTREME, I recall sending a few out.

**Comment by Asizon** [comment]
@hydra we are now new maintainers for the new Configurator maintaining system:
@haslinghuis
@limonspb
@chmelevskij
@Asizon

**Review by chmelevskij — APPROVED** [review]

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - Github identifies PR as mergeable -> FAIL
 - Assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - 'Don't merge' label NOT found -> PASS
 - At least one 'RN:' label found -> PASS
 - 'Tested' label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

**Review by schugabe — APPROVED** [review]
Flashing bf to a SPRacingH7RF works with this PR

File: src/js/protocols/stm32usbdfu.js
Comment (McGiverGim) [inline src/js/protocols/stm32usbdfu.js:352]: What is this? A bug in the FC? I don't understand how this works (the flash system). I suppose this cannot be fixed in any way and this is a workaround in the Configurator to support it?

Comment (McGiverGim) [reply src/js/protocols/stm32usbdfu.js:352]: I close this because the only change was the commit, so we are already supporting this.

**Review by McGiverGim — APPROVED** [review]

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - Github identifies PR as mergeable -> PASS
 - Assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - 'Don't merge' label NOT found -> PASS
 - At least one 'RN:' label found -> PASS
 - 'Tested' label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> PASS

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - Github identifies PR as mergeable -> FAIL
 - Assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - 'Don't merge' label NOT found -> PASS
 - At least one 'RN:' label found -> PASS
 - 'Tested' label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> PASS

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - Github identifies PR as mergeable -> PASS
 - Assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - 'Don't merge' label NOT found -> PASS
 - At least one 'RN:' label found -> PASS
 - 'Tested' label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> PASS

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> PASS
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> PASS

**Comment by hydra** [comment]
What additional testing is required in order to get the 'Tested' label added?

**Comment by haslinghuis** [comment]
@limonspb which hardware did you test with?
@hydra without an engineering sample we are not able to test this.

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> PASS
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> FAIL
 - approver count at least three -> PASS

**Comment by limonspb** [comment]
@haslinghuis  I haven't test anything. I saw @schugabe  tested, figured that would be enough. Apologize, I remove the tag.

**Comment by Asizon** [comment]
To make good quality merges the code should be tested by maintainers with the needed hardware.

**Comment by limonspb** [comment]
Gotcha!

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> PASS
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> PASS

**Comment by hydra** [comment]
if the current maintainers do not have the hardware then perhaps the old maintainers should send their hardware to them?

As there is a global shortage of CPUs there are no new FCs that can be sent out.

Is it not enough that the designer of the hardware and the person that initially added support for flashing to external flash chips, and other users, have tested it?  The updated code is isolated from non external flash chips too.

File: src/js/protocols/stm32usbdfu.js
Comment (mikeller) [inline src/js/protocols/stm32usbdfu.js:610]: What is the reason for matching this by size? This seems to be contrary to the way Intel Hex files are designed to work, and not really scalable / extensible:
- what happens if there are multiple partitions of the same size;
- what happens if the hex file is shorter than the partition it should go into;
- why is this not using the address, which is the way Intel Hex is meant to work.

Comment (hydra) [reply src/js/protocols/stm32usbdfu.js:610]: 1. Currently first matching partition is used, by design.  There are no targets like this anyway.
2. Short hex results in no match, intended.  All current EXST HEX images are the same size as the partition.
3. I agree!  The pattern of checking the size come from the non-external flash code.  To be honest the whole size checking should probably just be deleted since the next step of the dfu flashing processes deals with the actual addresses.  If it was up to me I would ditch .hex files and use a zipped .json file and improve the entire release and flashing process and allow multiple images/memory areas to be flashed/updated.  similar to how PX4 does it.  I'll be happy to delete the size checking code from both the internal and external flash use cases and make it rely on addresses only, however this would take some time and I've got other outstanding work right now.  Suggest merging this stop-gap solution until time allows for further improvements.

Comment (hydra) [reply src/js/protocols/stm32usbdfu.js:610]: With regard to 3 above.  This section can be deleted (mostly):
https://github.com/betaflight/betaflight-configurator/blob/26ec8fb0e241c385f7c2a0c67c4e4ff1a86b673a/src/js/protocols/stm32usbdfu.js#L572-L629

This section could could be improved (out of scope of this PR):

https://github.com/betaflight/betaflight-configurator/blob/26ec8fb0e241c385f7c2a0c67c4e4ff1a86b673a/src/js/protocols/stm32usbdfu.js#L777-L804

As you can see, all the addresses in the .hex are checked to see if they match an 'erasable page'.  Instead of checking the size which I fully agree is wrong for both internal an external flash chips as the hex can contain non-contiguous memory addresses, the code could be updated to show 'address not found' or 'address not writable' error messages instead.

**Comment by hydra** [comment]
I'm closing this PR.  I'll do another one that ditches the size checking for both internal and external flashes and relies on addresses alone which is a sensible long-term solution.

**Pull request closed without merging**
